core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -471,26 +482,33 @@ private static PartitionSpec updateSpecSchema(Schema schema, PartitionSpec parti
 
     // add all of the fields to the builder. IDs should not change.
     for (PartitionField field : partitionSpec.fields()) {
-      specBuilder.add(field.sourceId(), field.name(), field.transform().toString());
+      specBuilder.add(field.sourceId(), field.fieldId(), field.name(), field.transform().toString());
     }
 
     return specBuilder.build();
   }
 
-  private static PartitionSpec freshSpec(int specId, Schema schema, PartitionSpec partitionSpec) {
-    PartitionSpec.Builder specBuilder = PartitionSpec.builderFor(schema)
+  private static PartitionSpec freshSpecWithAssignIds(int specId, Schema newSchema, Schema oldSchema,
+                                                      PartitionSpec partitionSpec, AtomicInteger nextPartitionFieldId,
+                                                      Map<String, Integer> partitionFieldIdByColumnName) {
```

Comment:
I don't think that it is necessary to have this method here. The `freshSpec` method just ensures that all specs have the correct schema associated.